ocaml-lsp-server/src/ocaml_lsp_server.ml

```diff
@@ -683,7 +684,52 @@ let on_request :
       let change = { Lsp.Protocol.TextEdit.newText = result; range } in
       Ok (store, [ change ]) )
   | Lsp.Client_request.TextDocumentOnTypeFormatting _ -> Ok (store, [])
-  | Lsp.Client_request.SelectionRange _ -> Ok (store, [])
+  | Lsp.Client_request.SelectionRange { textDocument = { uri }; positions } ->
+    let selection_range_of_shapes (cursor_position : Lsp.Protocol.Position.t)
+        (shapes : Query_protocol.shape list) :
+        Lsp.Protocol.SelectionRange.t option =
+      let rec ranges_of_shape parent s =
+        let range = range_of_loc s.Query_protocol.shape_loc in
+        let selectionRange = { Lsp.Protocol.SelectionRange.range; parent } in
+        match s.Query_protocol.shape_sub with
+        | [] -> [ selectionRange ]
+        | xs -> List.concat_map xs ~f:(ranges_of_shape (Some selectionRange))
+      in
+      let ranges = List.concat_map ~f:(ranges_of_shape None) shapes in
+      (* note: `Inside whatever < `Outside whatever *)
+      let range_includes_pos r p =
+        let rs = r.Lsp.Protocol.Range.start_ in
+        let re = r.end_ in
+        if p < rs then
+          `Outside (abs (rs.line - p.line), abs (rs.character - p.character))
+        else if p > re then
+          `Outside (abs (re.line - p.line), abs (re.character - p.character))
+        else
+          `Inside
+            ( abs (rs.line - p.line) + abs (re.line - p.line)
+            , abs (rs.character - p.character) + abs (re.character - p.character)
+            )
+      in
+      (* try to find the nearest range inside first, then outside *)
+      let nearest_range =
+        List.map ranges ~f:(fun r ->
+            ( r
+            , range_includes_pos r.Lsp.Protocol.SelectionRange.range
+                cursor_position ))
+        |> List.sort ~compare:(fun r1 r2 ->
+               compare (snd r1) (snd r2) |> Ordering.of_int)
```

Comment:
You're comparing a polymorphic variant using polymorphic comparison. This is quite suspicious to me and we should write an explicit comparison function here.